project-base/storefront/components/Pages/Order/ContactInformation/contactInformationFormMeta.ts

```diff
@@ -56,55 +58,99 @@ export const useContactInformationForm = (): [UseFormReturn<ContactInformation>,
                 otherwise: Yup.string(),
             }),
-            deliveryStreet: Yup.string().when('differentDeliveryAddress', {
-                is: true,
+            deliveryStreet: Yup.string().when(['differentDeliveryAddress', 'deliveryAddressUuid'], {
+                is: (differentDeliveryAddress: boolean, deliveryAddressUuid: string) =>
+                    shouldValidateDeliveryAddressField(
+                        isUserLoggedIn,
+                        differentDeliveryAddress,
+                        deliveryAddressUuid,
+                        !!pickupPlace,
+                    ),
                 then: validateStreet(t),
                 otherwise: Yup.string(),
             }),
-            deliveryCity: Yup.string().when('differentDeliveryAddress', {
-                is: true,
+            deliveryCity: Yup.string().when(['differentDeliveryAddress', 'deliveryAddressUuid'], {
+                is: (differentDeliveryAddress: boolean, deliveryAddressUuid: string) =>
+                    shouldValidateDeliveryAddressField(
+                        isUserLoggedIn,
+                        differentDeliveryAddress,
+                        deliveryAddressUuid,
+                        !!pickupPlace,
+                    ),
                 then: validateCity(t),
                 otherwise: Yup.string(),
             }),
-            deliveryPostcode: Yup.string().when('differentDeliveryAddress', {
-                is: true,
+            deliveryPostcode: Yup.string().when(['differentDeliveryAddress', 'deliveryAddressUuid'], {
+                is: (differentDeliveryAddress: boolean, deliveryAddressUuid: string) =>
+                    shouldValidateDeliveryAddressField(
+                        isUserLoggedIn,
+                        differentDeliveryAddress,
+                        deliveryAddressUuid,
+                        !!pickupPlace,
+                    ),
                 then: validatePostcode(t),
                 otherwise: Yup.string(),
             }),
-
-            deliveryCountry: Yup.object().when('differentDeliveryAddress', {
-                is: true,
+            deliveryCountry: Yup.object().when(['differentDeliveryAddress', 'deliveryAddressUuid'], {
+                is: (differentDeliveryAddress: boolean, deliveryAddressUuid: string) =>
+                    shouldValidateDeliveryAddressField(
+                        isUserLoggedIn,
+                        differentDeliveryAddress,
+                        deliveryAddressUuid,
+                        !!pickupPlace,
+                    ),
                 then: validateCountry(t),
             }),
             newsletterSubscription: Yup.boolean(),
-            deliveryAddressUuid: Yup.string().optional().nullable(),
+            deliveryAddressUuid: Yup.string().optional(),
             note: Yup.string().optional().nullable(),
         }),
     );
-    const defaultValues = contactInformationValues;
+    const defaultValues: ContactInformation = {
```

Comment:
What about proposal which Ivan had about put typing to those form hooks? Cannot we implement them with this occasion? This way we would already know the type from the `contactInformationValues` and not need to type it manually.

Reply:
The type there is kind of unnecessary. Even without it, it works correctly (adding or removing property breaks TS). The only reason I added it is to see this:

<img width="657" alt="image" src="https://github.com/shopsys/shopsys/assets/55181354/c2d20eb4-8706-4988-b81b-7041df2880dd">

Instead of this:

<img width="657" alt="image" src="https://github.com/shopsys/shopsys/assets/55181354/94ecbf83-fcdf-4034-840b-e2eac9373d22">

But maybe that is unnecessary. I removed the type completely, as it is redundant, but if you have a different suggestion, then let me know 😇